File: Modules/SekaiMasterUpdater/master.py
import orjson
import asyncio
import aiofiles
from pathlib import Path
from aiopath import AsyncPath
from typing import Dict, List, Union, Optional
from aiohttp import ClientSession, ClientResponse

from .compare import compare_version
from ..logger import AsyncLogger
from ..SekaiClient.manager import SekaiClientManager
from ..SekaiClient.model import SekaiServerRegion, SekaiServerInfo, SekaiApiHttpStatus, HarukiAssetUpdaterInfo

logger = AsyncLogger(__name__, level="DEBUG")


class SekaiMasterUpdater:
    def __init__(
        self,
        servers: Dict[SekaiServerRegion, SekaiServerInfo],
        managers: Dict[SekaiServerRegion, SekaiClientManager],
        master_dirs: Dict[SekaiServerRegion, Union[Path, str]],
        version_dirs: Dict[SekaiServerRegion, Union[Path, str]],
        asset_updater_servers: List[HarukiAssetUpdaterInfo],
    ) -> None:
        self.servers = servers
        self.managers = managers
        self.master_dirs = master_dirs
        self.version_dirs = version_dirs
        self.asset_updater_servers = asset_updater_servers

    @staticmethod
    async def load_file(file_path: Union[AsyncPath, Path, str]) -> Union[Dict, List]:
        async with aiofiles.open(file_path, "r", encoding="utf-8") as file:
            return orjson.loads(await file.read())

    @staticmethod
    async def save_file(file_path: Union[AsyncPath, Path, str], data: Union[Dict, List]) -> None:
        async with aiofiles.open(file_path, "wb") as file:
            await file.write(orjson.dumps(data, option=orjson.OPT_INDENT_2))

    # Call Haruki Sekai Asset Updater
    async def _call_asset_updater(self, options: Dict) -> Optional[ClientResponse]:
        async with ClientSession() as session:
            async with session.request(**options) as response:
                if response.status == SekaiApiHttpStatus.OK:
                    return response
                elif response.status == SekaiApiHttpStatus.CONFLICT:
                    await asyncio.sleep(60)
                    await self._call_asset_updater(options)
                    return None
                else:
                    return None

    async def asset_updater(self, server: SekaiServerRegion, data: Dict):
        body = {
            "server": server.value,
            "assetVersion": data.get("assetVersion"),
            "assetHash": data.get("assetHash", None),
        }
        for updater in self.asset_updater_servers:
            options = {
                "url": f"{updater.url}/update_asset",
                "method": "POST",
                "json": body,
                "headers": {"User-Agent": "Haruki Sekai API/v3.0.0"},
            }
            if updater.authorization:
                options["headers"]["Authorization"] = f"Bearer {updater.authorization}"
            asyncio.create_task(self._call_asset_updater(options))

    # Check update
    async def check_update(self, server: SekaiServerRegion, manager: SekaiClientManager) -> Optional[Dict]:
        _update_master = False
        _update_asset = False

        version_file = AsyncPath(self.version_dirs.get(server)) / "current_version.json"
        current_version = await self.load_file(version_file)
        current_server_version = await manager.get_login_data()
        current_data_version = current_version.get("dataVersion")
        current_asset_version = current_version.get("assetVersion")
        current_server_data_version = current_server_version.get("dataVersion")
        current_server_asset_version = current_server_version.get("assetVersion")
        current_server_asset_hash = current_server_version.get("assetHash", None)

        if server in [SekaiServerRegion.JP, SekaiServerRegion.EN]:
            if await asyncio.to_thread(compare_version, current_server_data_version, current_data_version):
                await logger.critical(
                    f"{server.value.upper()} server found new master data version: {current_server_data_version}"
                )
                _update_master = True
            if await asyncio.to_thread(compare_version, current_server_asset_version, current_asset_version):
                await logger.critical(
                    f"{server.value.upper()} server found new asset version: {current_server_data_version}"
                )
                _update_asset = True
        else:
            current_cdn_version = current_version.get("cdnVersion")
            current_server_cdn_version = current_server_version.get("cdnVersion")
            if int(current_cdn_version) < int(current_server_cdn_version):
                await logger.critical(
                    f"{server.value.upper()} server found new cdn version: {current_server_cdn_version}"
                )
                _update_master = True
                _update_asset = True

        if _update_asset:
            await self.asset_updater(server, current_server_version)
        if _update_master:
            await self.update_master(server, manager)

        if _update_asset or _update_master:
            current_version["dataVersion"] = current_server_data_version
            current_version["assetVersion"] = current_server_asset_version
            current_version["assetHash"] = current_server_asset_hash
            if current_server_version.get("cdnVersion"):
                current_version["cdnVersion"] = current_server_version.get("cdnVersion")
            await self.save_file(version_file, current_version)
            await self.save_file(
                AsyncPath(self.version_dirs.get(server)) / f"{current_server_data_version}.json", current_version
            )
            return {server: current_server_data_version}
        return None

    # Check update coroutine
    async def check_update_concurrently(self) -> Optional[Dict]:
        await logger.start()
        result_dict = {}
        tasks = [self.check_update(server, manager) for server, manager in self.managers.items()]
        results = await asyncio.gather(*tasks)
        result_dict.update(
            (key, value) for result in results if result and isinstance(result, dict) for key, value in result.items()
        )
        if result_dict:
            return result_dict
        await logger.stop()
        return None

    async def _save_split_master_data(self, server: SekaiServerRegion, master: Dict) -> None:
        await logger.info(
            f"Saving {server.value.upper()} server split master data...",
        )
        _master_dir = AsyncPath(self.master_dirs.get(server))
        tasks = [self.save_file(_master_dir / f"{key}.json", value) for key, value in master.items()]
        await asyncio.gather(*tasks)
        await logger.info(f"Saved {server.value.upper()} server split master data.")

    async def update_master(self, server: SekaiServerRegion, manager: SekaiClientManager) -> None:
        await logger.info(f"Downloading {server.value.upper()} new master data...")
        master_data = await manager.download_master()
        await logger.info(f"Downloaded {server.value.upper()} new master data.")
        await self._save_split_master_data(server, master_data)

File: updater/master.go
package updater

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"haruki-sekai-api/client"
	"haruki-sekai-api/utils"
	"haruki-sekai-api/utils/logger"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"sync"
	"time"

	"github.com/bytedance/sonic"
)

type SekaiMasterUpdater struct {
	servers             map[utils.HarukiSekaiServerRegion]client.SekaiServerInfo
	managers            map[utils.HarukiSekaiServerRegion]*client.SekaiClientManager
	masterDirs          map[utils.HarukiSekaiServerRegion]string
	versionDirs         map[utils.HarukiSekaiServerRegion]string
	assetUpdaterServers []client.HarukiAssetUpdaterInfo
	logger              *logger.Logger
}

func NewSekaiMasterUpdater(
	servers map[utils.HarukiSekaiServerRegion]client.SekaiServerInfo,
	managers map[utils.HarukiSekaiServerRegion]*client.SekaiClientManager,
	masterDirs map[utils.HarukiSekaiServerRegion]string,
	versionDirs map[utils.HarukiSekaiServerRegion]string,
	assetUpdaterServers []client.HarukiAssetUpdaterInfo,
) *SekaiMasterUpdater {
	return &SekaiMasterUpdater{
		servers:             servers,
		managers:            managers,
		masterDirs:          masterDirs,
		versionDirs:         versionDirs,
		assetUpdaterServers: assetUpdaterServers,
		logger:              logger.NewLogger("SekaiMasterUpdater", "DEBUG", nil),
	}
}

func (s *SekaiMasterUpdater) loadFile(filePath string) (map[string]any, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	var result map[string]any
	if err := sonic.Unmarshal(data, &result); err != nil {
		return nil, err
	}
	return result, nil
}

func (s *SekaiMasterUpdater) saveFile(filePath string, data any) error {
	jsonData, err := sonic.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}

	// 确保目录存在
	dir := filepath.Dir(filePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}

	return os.WriteFile(filePath, jsonData, 0644)
}

// callAssetUpdater 调用 Haruki Sekai Asset Updater
func (s *SekaiMasterUpdater) callAssetUpdater(ctx context.Context, options map[string]any) (*http.Response, error) {
	cli := &http.Client{Timeout: 30 * time.Second}

	url, ok := options["url"].(string)
	if !ok {
		return nil, fmt.Errorf("invalid url")
	}

	method, ok := options["method"].(string)
	if !ok {
		return nil, fmt.Errorf("invalid method")
	}

	jsonBody := options["json"]
	bodyBytes, err := json.Marshal(jsonBody)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewBuffer(bodyBytes))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")

	if headers, ok := options["headers"].(map[string]string); ok {
		for k, v := range headers {
			req.Header.Set(k, v)
		}
	}

	resp, err := cli.Do(req)
	if err != nil {
		return nil, err
	}

	switch resp.StatusCode {
	case 200: // SekaiApiHttpStatusOk
		return resp, nil
	case 409: // SekaiApiHttpStatusConflict
		resp.Body.Close()
		time.Sleep(60 * time.Second)
		return s.callAssetUpdater(ctx, options)
	default:
		resp.Body.Close()
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}
}

func (s *SekaiMasterUpdater) assetUpdater(ctx context.Context, server utils.HarukiSekaiServerRegion, data map[string]any) {
	body := map[string]any{
		"server": string(server),
	}

	if assetVersion, ok := data["assetVersion"]; ok {
		body["assetVersion"] = assetVersion
	}

	if assetHash, ok := data["assetHash"]; ok {
		body["assetHash"] = assetHash
	}

	var wg sync.WaitGroup
	for _, updater := range s.assetUpdaterServers {
		wg.Add(1)
		go func(upd client.HarukiAssetUpdaterInfo) {
			defer wg.Done()

			options := map[string]any{
				"url":    upd.Url + "/update_asset",
				"method": "POST",
				"json":   body,
				"headers": map[string]string{
					"User-Agent": "Haruki Sekai API/v3.0.0",
				},
			}

			if upd.Authorization != "" {
				headers := options["headers"].(map[string]string)
				headers["Authorization"] = "Bearer " + upd.Authorization
			}

			resp, err := s.callAssetUpdater(ctx, options)
			if err != nil {
				s.logger.Errorf("Failed to call asset updater: %v", err)
				return
			}
			if resp != nil {
				resp.Body.Close()
			}
		}(updater)
	}
	wg.Wait()
}

// checkUpdate 检查更新
func (s *SekaiMasterUpdater) checkUpdate(ctx context.Context, server utils.HarukiSekaiServerRegion, manager *client.SekaiClientManager) (map[utils.HarukiSekaiServerRegion]string, error) {
	updateMaster := false
	updateAsset := false

	versionFile := filepath.Join(s.versionDirs[server], "current_version.json")
	currentVersion, err := s.loadFile(versionFile)
	if err != nil {
		return nil, fmt.Errorf("failed to load version file: %v", err)
	}

	currentServerVersion, err := manager.GetLoginData()
	if err != nil {
		return nil, fmt.Errorf("failed to get login data: %v", err)
	}

	currentDataVersion := getString(currentVersion, "dataVersion")
	currentAssetVersion := getString(currentVersion, "assetVersion")
	currentServerDataVersion := getString(currentServerVersion, "dataVersion")
	currentServerAssetVersion := getString(currentServerVersion, "assetVersion")
	currentServerAssetHash := getString(currentServerVersion, "assetHash")

	if server == utils.HarukiSekaiServerRegionJP || server == utils.HarukiSekaiServerRegionEN {
		// 检查数据版本更新
		isNewer, err := utils.CompareVersion(currentServerDataVersion, currentDataVersion)
		if err != nil {
			s.logger.Warnf("Failed to compare data version: %v", err)
		} else if isNewer {
			s.logger.Criticalf("%s server found new master data version: %s",
				string(server), currentServerDataVersion)
			updateMaster = true
		}

		// 检查资源版本更新
		isNewer, err = utils.CompareVersion(currentServerAssetVersion, currentAssetVersion)
		if err != nil {
			s.logger.Warnf("Failed to compare asset version: %v", err)
		} else if isNewer {
			s.logger.Criticalf("%s server found new asset version: %s",
				string(server), currentServerAssetVersion)
			updateAsset = true
		}
	} else {
		// 对于其他服务器，检查 CDN 版本
		currentCdnVersion := getInt(currentVersion, "cdnVersion")
		currentServerCdnVersion := getInt(currentServerVersion, "cdnVersion")

		if currentCdnVersion < currentServerCdnVersion {
			s.logger.Criticalf("%s server found new cdn version: %d",
				string(server), currentServerCdnVersion)
			updateMaster = true
			updateAsset = true
		}
	}

	if updateAsset {
		s.assetUpdater(ctx, server, currentServerVersion)
	}

	if updateMaster {
		if err := s.updateMaster(ctx, server, manager); err != nil {
			return nil, fmt.Errorf("failed to update master: %v", err)
		}
	}

	if updateAsset || updateMaster {
		// 更新版本文件
		currentVersion["dataVersion"] = currentServerDataVersion
		currentVersion["assetVersion"] = currentServerAssetVersion
		currentVersion["assetHash"] = currentServerAssetHash

		if cdnVersion, ok := currentServerVersion["cdnVersion"]; ok {
			currentVersion["cdnVersion"] = cdnVersion
		}

		if err := s.saveFile(versionFile, currentVersion); err != nil {
			return nil, fmt.Errorf("failed to save version file: %v", err)
		}

		versionedFile := filepath.Join(s.versionDirs[server], currentServerDataVersion+".json")
		if err := s.saveFile(versionedFile, currentVersion); err != nil {
			return nil, fmt.Errorf("failed to save versioned file: %v", err)
		}

		return map[utils.HarukiSekaiServerRegion]string{server: currentServerDataVersion}, nil
	}

	return nil, nil
}

// checkUpdateConcurrently 并发检查更新
func (s *SekaiMasterUpdater) CheckUpdateConcurrently(ctx context.Context) (map[utils.HarukiSekaiServerRegion]string, error) {
	s.logger.Infof("Starting concurrent update check...")

	type result struct {
		data map[utils.HarukiSekaiServerRegion]string
		err  error
	}

	resultChan := make(chan result, len(s.managers))
	var wg sync.WaitGroup

	for server, manager := range s.managers {
		wg.Add(1)
		go func(srv utils.HarukiSekaiServerRegion, mgr *client.SekaiClientManager) {
			defer wg.Done()
			data, err := s.checkUpdate(ctx, srv, mgr)
			resultChan <- result{data: data, err: err}
		}(server, manager)
	}

	wg.Wait()
	close(resultChan)

	resultDict := make(map[utils.HarukiSekaiServerRegion]string)
	for res := range resultChan {
		if res.err != nil {
			s.logger.Errorf("Update check error: %v", res.err)
			continue
		}
		if res.data != nil {
			for k, v := range res.data {
				resultDict[k] = v
			}
		}
	}

	if len(resultDict) > 0 {
		return resultDict, nil
	}

	s.logger.Infof("Update check completed, no updates found")
	return nil, nil
}

func (s *SekaiMasterUpdater) saveSplitMasterData(server utils.HarukiSekaiServerRegion, master map[string]any) error {
	s.logger.Infof("Saving %s server split master data...", string(server))

	masterDir := s.masterDirs[server]
	if err := os.MkdirAll(masterDir, 0755); err != nil {
		return fmt.Errorf("failed to create master directory: %v", err)
	}

	var wg sync.WaitGroup
	errChan := make(chan error, len(master))

	for key, value := range master {
		wg.Add(1)
		go func(k string, v any) {
			defer wg.Done()
			filePath := filepath.Join(masterDir, k+".json")
			if err := s.saveFile(filePath, v); err != nil {
				errChan <- fmt.Errorf("failed to save %s: %v", k, err)
			}
		}(key, value)
	}

	wg.Wait()
	close(errChan)

	for err := range errChan {
		if err != nil {
			return err
		}
	}

	s.logger.Infof("Saved %s server split master data", string(server))
	return nil
}

func (s *SekaiMasterUpdater) updateMaster(ctx context.Context, server utils.HarukiSekaiServerRegion, manager *client.SekaiClientManager) error {
	s.logger.Infof("Downloading %s new master data...", string(server))

	masterData, err := manager.DownloadMaster()
	if err != nil {
		return fmt.Errorf("failed to download master data: %v", err)
	}

	s.logger.Infof("Downloaded %s new master data", string(server))

	return s.saveSplitMasterData(server, masterData)
}

// 辅助函数
func getString(m map[string]any, key string) string {
	if v, ok := m[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

func getInt(m map[string]any, key string) int {
	if v, ok := m[key]; ok {
		switch val := v.(type) {
		case int:
			return val
		case float64:
			return int(val)
		case string:
			if i, err := strconv.Atoi(val); err == nil {
				return i
			}
		}
	}
	return 0
}
